File: cotsSource/org.apache.qpid/ruby/lib/qpid/codec.rb
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

require 'qpid/packer.rb'
require 'iconv'

module Qpid

  class Codec

    include Qpid::Packer

    attr_reader :spec
    
    def initialize(spec = "")
      @spec = spec
    end

    def write_void(v)
      unless v.nil?
        raise Exception.new("void not nil: #{v}")
      end
    end

    def read_void
      return nil
    end

    def write_bit(b)
      unless b
        raise Exception.new("bit is nil: #{b}")
      end
    end

    def read_bit
      return true
    end

    def read_uint8
      return unpack("C", 1)
    end

    def write_uint8(n)
      return pack("C", n)
    end

    def read_int8
      return unpack("c", 1)
    end

    def write_int8(n)
      pack("c", n)
    end

    def read_char
      return unpack("c", 1)
    end

    def write_char(c)
      pack("c")
    end

    def read_boolean
      return read_uint8 != 0
    end

    def write_boolean(b)
      n = 0
      n = 1 if b != 0
      write_uint8(n)
    end

    def read_uint16
      return unpack("n", 2)
    end

    def write_uint16(n)
      pack("n", n)
    end

    def read_int16
      # XXX: holy moly.. pack/unpack doesn't have signed network byte order.  Crazy hackery.
      val = unpack("n", 2)
      val -= 2 ** 16  if val >= 2 ** 15
      return val
    end

    def write_int16(n)
      # XXX: Magically this one works even though it's not signed.
      pack("n", n)
    end

    def read_uint32
      return unpack("N", 4)
    end

    def write_uint32(n)
      pack("N", n)
    end

    def read_int32
      # Again no pack/unpack for signed int
      return unpack("N", 4)
    end

    def write_int32(n)
      # FIXME
      pack("N", n)
    end

    def read_float
      return unpack("g", 4)
    end

    def write_float(n)
      pack("g", n)
    end

    def read_sequence_no
      return read_uint32.to_serial
    end

    def write_sequence_no(n)
      write_uint32(n.value)
    end

    def encode_64bit(num, signed = false)
      b = []

      if num < 0 && signed
        num += 2 ** 64
      end

      (0..7).each do |c|
        d = 7 - c
        b[c] = (num & (0xff << d * 8)) >> d * 8
      end
      pack('C8', *b)
    end


    def decode_64bit(signed = false)
      # Silly ruby pack/unpack does not implement 64 bit network byte order
      # encode/decode.
      a = unpack('C8', 8)
      num = 0
      (0..7).each do |c|
        d = 7 - c
        num |= a[c] << 8 * d
      end

      if signed && num >= 2 ** 63
        num -= 2 ** 64
      end
      return num
    end

    def read_uint64
      return decode_64bit
    end

    def write_uint64(n)
      encode_64bit(n)
    end

    def read_int64
      return decode_64bit(signed = true)
    end

    def write_int64(n)
      encode_64bit(n, signed = true)
    end

    def read_datetime
      return read_uint64
    end

    def write_datetime(n)
      write_uint64(n)
    end

    def read_double
      return unpack("G", 8)
    end

    def write_double(n)
      pack("G", n)
    end

    def read_vbin8
      # XXX
      return read(read_uint8)
    end

    def write_vbin8(b)
      # XXX
      write_uint8(b.length)
      write(b)
    end

    def read_str8
      # FIXME: Check iconv.. I think this will throw if there are odd characters.
      return Iconv.conv("ASCII", "UTF-8", read_vbin8)
    end

    def write_str8(s)
      write_vbin8(Iconv.conv("UTF-8", "ASCII", s))
    end

    def read_str16
      return Iconv.conv("ASCII", "UTF-8", read_vbin16)
    end

    def write_str16(s)
      write_vbin16(Iconv.conv("UTF-8", "ASCII", s))
    end

    def read_vbin16
      # XXX: Using read method?
      return read(read_uint16)
    end

    def write_vbin16(b)
      write_uint16(b.length)
      write(b)
    end

    def read_sequence_set
      # FIXME: Need datatypes
      result = RangedSet.new
      size = read_uint16
      nranges = size / 8
      nranges.times do |i|
        lower = read_sequence_no
        upper = read_sequence_no
        result.add(lower, upper)
      end
      return result
    end

    def write_sequence_set(ss)
      size = 8 * ss.ranges.length
      write_uint16(size)
      ss.ranges.each do |range|
        write_sequence_no(range.lower)
        write_sequence_no(range.upper)
      end
    end

    def read_vbin32
      return read(read_uint32)
    end

    def write_vbin32(b)
      write_uint32(b.length)
      write(b)
    end

    def write_map(m)
      sc = StringCodec.new(@spec)
      unless m.nil?
        sc.write_uint32(m.size)
        m.each do |k, v|
          unless type = @spec.encoding(v.class)
            raise Exception.new("no encoding for: #{v.class}")
          end
          sc.write_str8(k)
          sc.write_uint8(type.code)
          type.encode(sc, v)
        end
      end
      write_vbin32(sc.encoded)
    end

    def read_map
      sc = StringCodec.new(@spec, read_vbin32)
      return nil unless sc.encoded
      count = sc.read_uint32
      result = nil
      if count
        result = {}
        until sc.encoded.empty?
          k = sc.read_str8
          code = sc.read_uint8
          type = @spec.types[code]
          v = type.decode(sc)
          result[k] = v
        end
      end
      return result
    end

    def write_array(a)
      sc = StringCodec.new(@spec)
      unless a.nil?
        if a.length > 0
          type = @spec.encoding(a[0].class)
        else
          type = @spec.encoding(nil.class)
        end
        sc.write_uint8(type.code)
        sc.write_uint32(a.size)
        a.each { |o| type.encode(sc, o) }
      end
      write_vbin32(sc.encoded)
    end

    def read_array
      sc = StringCodec.new(@spec, read_vbin32)
      return nil if not sc.encoded
      type = @spec.types[sc.read_uint8]
      count = sc.read_uint32
      result = nil
      if count
        result = []
        count.times { |i| result << (type.decode(sc)) }
      end
      return result
    end

    def write_list(l)
      sc = StringCodec.new(@spec)
      unless l.nil?
        sc.write_uint32(l.length)
        l.each do |o|
          type = @spec.encoding(o.class)
          sc.write_uint8(type.code)
          type.encode(sc, o)
        end
      end
      write_vbin32(sc.encoded)
    end

    def read_list
      sc = StringCodec.new(@spec, read_vbin32)
      return nil if not sc.encoded
      count = sc.read_uint32
      result = nil
      if count
        result = []
        count.times do |i|
          type = @spec.types[sc.read_uint8]
          result << type.decode(sc)
        end
      end
      return result
    end

    def read_struct32
      size = read_uint32
      code = read_uint16
      type = @spec.structs[code]
      # XXX: BLEH!
      fields = type.decode_fields(self)
      return Qpid::struct(type, fields)
    end

    def write_struct32(value)
      type = value.st_type
      sc = StringCodec.new(@spec)
      sc.write_uint16(type.code)
      type.encode_fields(sc, value)
      write_vbin32(sc.encoded)
    end

    def read_control
      cntrl = @spec.controls[read_uint16]
      return Qpid::struct(cntrl, cntrl.decode_fields(self))
    end

    def write_control(ctrl)
      type = ctrl.st_type
      write_uint16(type.code)
      type.encode_fields(self, ctrl)
    end

    def read_command
      type = @spec.commands[read_uint16]
      hdr = @spec[:header].decode(self)
      cmd = Qpid::struct(type, type.decode_fields(self))
      return hdr, cmd
    end

    def write_command(hdr, cmd)
      type = cmd.st_type
      write_uint16(type.code)
      hdr.st_type.encode(self, hdr)
      type.encode_fields(self, cmd)
    end

    def read_size(width)
      if width > 0
        return send(:"read_uint#{width * 8}")
      end
    end

    def write_size(width, n)
      if width > 0
        send(:"write_uint#{width * 8}", n)
      end
    end

    def read_uuid
      return unpack("a16", 16)
    end

    def write_uuid(s)
      pack("a16", s)
    end

    def read_bin128
      return unpack("a16", 16)
    end

    def write_bin128(b)
      pack("a16", b)
    end

  end

  class StringCodec < Codec

    def initialize(spec, encoded = "")
      @spec = spec
      @encoded = encoded
    end

    attr_reader :encoded

    def write(s)
      @encoded += s
    end

    def read(n)
      return "" if n.nil?
      result = @encoded[0...n]
      @encoded = @encoded[n...@encoded.size] || ""
      return result
    end
  end
end

File: cotsSource/org.apache.qpid/python/qpid/codec.py
#!/usr/bin/env python

#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

"""
Utility code to translate between python objects and AMQP encoded data
fields.

The unit test for this module is located in tests/codec.py
"""

import re, qpid, spec08
from cStringIO import StringIO
from struct import *
from reference import ReferenceId

class EOF(Exception):
  pass

TYPE_ALIASES = {
  "long_string": "longstr",
  "unsigned_int": "long"
  }

class Codec:

  """
  class that handles encoding/decoding of AMQP primitives
  """

  def __init__(self, stream, spec):
    """
    initializing the stream/fields used
    """
    self.stream = stream
    self.spec = spec
    self.nwrote = 0
    self.nread = 0
    self.incoming_bits = []
    self.outgoing_bits = []

    self.types = {}
    self.codes = {}
    self.encodings = {
      basestring: "longstr",
      int: "long",
      long: "long",
      None.__class__:"void",
      list: "sequence",
      tuple: "sequence",
      dict: "table"
      }

    for constant in self.spec.constants:
      if constant.klass == "field-table-type":
        type = constant.name.replace("field_table_", "")
        self.typecode(constant.id, TYPE_ALIASES.get(type, type))

    if not self.types:
      self.typecode(ord('S'), "longstr")
      self.typecode(ord('I'), "long")

  def typecode(self, code, type):
    self.types[code] = type
    self.codes[type] = code

  def resolve(self, klass):
    if self.encodings.has_key(klass):
      return self.encodings[klass]
    for base in klass.__bases__:
      result = self.resolve(base)
      if result != None:
        return result

  def read(self, n):
    """
    reads in 'n' bytes from the stream. Can raise EOF exception
    """
    self.clearbits()
    data = self.stream.read(n)
    if n > 0 and len(data) == 0:
      raise EOF()
    self.nread += len(data)
    return data

  def write(self, s):
    """
    writes data 's' to the stream
    """
    self.flushbits()
    self.stream.write(s)
    self.nwrote += len(s)

  def flush(self):
    """
    flushes the bits and data present in the stream
    """
    self.flushbits()
    self.stream.flush()

  def flushbits(self):
    """
    flushes the bits(compressed into octets) onto the stream
    """
    if len(self.outgoing_bits) > 0:
      bytes = []
      index = 0
      for b in self.outgoing_bits:
        if index == 0: bytes.append(0)
        if b: bytes[-1] |= 1 << index
        index = (index + 1) % 8
      del self.outgoing_bits[:]
      for byte in bytes:
        self.encode_octet(byte)

  def clearbits(self):
    if self.incoming_bits:
      self.incoming_bits = []

  def pack(self, fmt, *args):
    """
    packs the data 'args' as per the format 'fmt' and writes it to the stream
    """
    self.write(pack(fmt, *args))

  def unpack(self, fmt):
    """
    reads data from the stream and unpacks it as per the format 'fmt'
    """
    size = calcsize(fmt)
    data = self.read(size)
    values = unpack(fmt, data)
    if len(values) == 1:
      return values[0]
    else:
      return values

  def encode(self, type, value):
    """
    calls the appropriate encode function e.g. encode_octet, encode_short etc.
    """
    if isinstance(type, spec08.Struct):
      self.encode_struct(type, value)
    else:
      getattr(self, "encode_" + type)(value)

  def decode(self, type):
    """
    calls the appropriate decode function e.g. decode_octet, decode_short etc.
    """
    if isinstance(type, spec08.Struct):
      return self.decode_struct(type)
    else:
      return getattr(self, "decode_" + type)()

  def encode_bit(self, o):
    """
    encodes a bit
    """
    if o:
      self.outgoing_bits.append(True)
    else:
      self.outgoing_bits.append(False)

  def decode_bit(self):
    """
    decodes a bit
    """
    if len(self.incoming_bits) == 0:
      bits = self.decode_octet()
      for i in range(8):
        self.incoming_bits.append(bits >> i & 1 != 0)
    return self.incoming_bits.pop(0)

  def encode_octet(self, o):
    """
    encodes octet (8 bits) data 'o' in network byte order
    """

    # octet's valid range is [0,255]
    if (o < 0 or o > 255):
        raise ValueError('Valid range of octet is [0,255]')

    self.pack("!B", int(o))

  def decode_octet(self):
    """
    decodes a octet (8 bits) encoded in network byte order
    """
    return self.unpack("!B")

  def encode_short(self, o):
    """
    encodes short (16 bits) data 'o' in network byte order
    """

    # short int's valid range is [0,65535]
    if (o < 0 or o > 65535):
        raise ValueError('Valid range of short int is [0,65535]: %s' % o)

    self.pack("!H", int(o))

  def decode_short(self):
    """
    decodes a short (16 bits) in network byte order
    """
    return self.unpack("!H")

  def encode_long(self, o):
    """
    encodes long (32 bits) data 'o' in network byte order
    """

    # we need to check both bounds because on 64 bit platforms
    # struct.pack won't raise an error if o is too large
    if (o < 0 or o > 4294967295):
      raise ValueError('Valid range of long int is [0,4294967295]')

    self.pack("!L", int(o))

  def decode_long(self):
    """
    decodes a long (32 bits) in network byte order
    """
    return self.unpack("!L")

  def encode_signed_long(self, o):
    self.pack("!q", o)

  def decode_signed_long(self):
    return self.unpack("!q")

  def encode_signed_int(self, o):
    self.pack("!l", o)

  def decode_signed_int(self):
    return self.unpack("!l")

  def encode_longlong(self, o):
    """
    encodes long long (64 bits) data 'o' in network byte order
    """
    self.pack("!Q", o)

  def decode_longlong(self):
    """
    decodes a long long (64 bits) in network byte order
    """
    return self.unpack("!Q")

  def encode_float(self, o):
    self.pack("!f", o)

  def decode_float(self):
    return self.unpack("!f")

  def encode_double(self, o):
    self.pack("!d", o)

  def decode_double(self):
    return self.unpack("!d")

  def encode_bin128(self, b):
    for idx in range (0,16):
      self.pack("!B", ord (b[idx]))

  def decode_bin128(self):
    result = ""
    for idx in range (0,16):
      result = result + chr (self.unpack("!B"))
    return result

  def encode_raw(self, len, b):
    for idx in range (0,len):
      self.pack("!B", b[idx])

  def decode_raw(self, len):
    result = ""
    for idx in range (0,len):
      result = result + chr (self.unpack("!B"))
    return result

  def enc_str(self, fmt, s):
    """
    encodes a string 's' in network byte order as per format 'fmt'
    """
    size = len(s)
    self.pack(fmt, size)
    self.write(s)

  def dec_str(self, fmt):
    """
    decodes a string in network byte order as per format 'fmt'
    """
    size = self.unpack(fmt)
    return self.read(size)

  def encode_shortstr(self, s):
    """
    encodes a short string 's' in network byte order
    """

    # short strings are limited to 255 octets
    if len(s) > 255:
        raise ValueError('Short strings are limited to 255 octets')

    self.enc_str("!B", s)

  def decode_shortstr(self):
    """
    decodes a short string in network byte order
    """
    return self.dec_str("!B")

  def encode_longstr(self, s):
    """
    encodes a long string 's' in network byte order
    """
    if isinstance(s, dict):
      self.encode_table(s)
    else:
      self.enc_str("!L", s)

  def decode_longstr(self):
    """
    decodes a long string 's' in network byte order
    """
    return self.dec_str("!L")

  def encode_table(self, tbl):
    """
    encodes a table data structure in network byte order
    """
    enc = StringIO()
    codec = Codec(enc, self.spec)
    if tbl:
      for key, value in tbl.items():
        if self.spec.major == 8 and self.spec.minor == 0 and len(key) > 128:
          raise ValueError("field table key too long: '%s'" % key)
        type = self.resolve(value.__class__)
        if type == None:
          raise ValueError("no encoding for: " + value.__class__)
        codec.encode_shortstr(key)
        codec.encode_octet(self.codes[type])
        codec.encode(type, value)
    s = enc.getvalue()
    self.encode_long(len(s))
    self.write(s)

  def decode_table(self):
    """
    decodes a table data structure in network byte order
    """
    size = self.decode_long()
    start = self.nread
    result = {}
    while self.nread - start < size:
      key = self.decode_shortstr()
      code = self.decode_octet()
      if self.types.has_key(code):
        value = self.decode(self.types[code])
      else:
        w = width(code)
        if fixed(code):
          value = self.read(w)
        else:
          value = self.read(self.dec_num(w))
      result[key] = value
    return result

  def encode_timestamp(self, t):
    """
    encodes a timestamp data structure in network byte order
    """
    self.encode_longlong(t)

  def decode_timestamp(self):
    """
    decodes a timestamp data structure in network byte order
    """
    return self.decode_longlong()

  def encode_content(self, s):
    """
    encodes a content data structure in network byte order

    content can be passed as a string in which case it is assumed to
    be inline data, or as an instance of ReferenceId indicating it is
    a reference id
    """
    if isinstance(s, ReferenceId):
      self.encode_octet(1)
      self.encode_longstr(s.id)
    else:
      self.encode_octet(0)
      self.encode_longstr(s)

  def decode_content(self):
    """
    decodes a content data structure in network byte order

    return a string for inline data and a ReferenceId instance for
    references
    """
    type = self.decode_octet()
    if type == 0:
      return self.decode_longstr()
    else:
      return ReferenceId(self.decode_longstr())

  # new domains for 0-10:

  def encode_rfc1982_long(self, s):
    self.encode_long(s)

  def decode_rfc1982_long(self):
    return self.decode_long()

  def encode_rfc1982_long_set(self, s):
    self.encode_short(len(s) * 4)
    for i in s:
      self.encode_long(i)

  def decode_rfc1982_long_set(self):
    count = self.decode_short() / 4
    set = []
    for i in range(0, count):
      set.append(self.decode_long())
    return set;

  def encode_uuid(self, s):
    self.pack("16s", s)

  def decode_uuid(self):
    return self.unpack("16s")

  def enc_num(self, width, n):
    if width == 1:
      self.encode_octet(n)
    elif width == 2:
      self.encode_short(n)
    elif width == 3:
      self.encode_long(n)
    else:
      raise ValueError("invalid width: %s" % width)

  def dec_num(self, width):
    if width == 1:
      return self.decode_octet()
    elif width == 2:
      return self.decode_short()
    elif width == 4:
      return self.decode_long()
    else:
      raise ValueError("invalid width: %s" % width)

  def encode_struct(self, type, s):
    if type.size:
      enc = StringIO()
      codec = Codec(enc, self.spec)
      codec.encode_struct_body(type, s)
      codec.flush()
      body = enc.getvalue()
      self.enc_num(type.size, len(body))
      self.write(body)
    else:
      self.encode_struct_body(type, s)

  def decode_struct(self, type):
    if type.size:
      size = self.dec_num(type.size)
      if size == 0:
        return None
    return self.decode_struct_body(type)

  def encode_struct_body(self, type, s):
    reserved = 8*type.pack - len(type.fields)
    assert reserved >= 0

    for f in type.fields:
      if s == None:
        self.encode_bit(False)
      elif f.type == "bit":
        self.encode_bit(s.get(f.name))
      else:
        self.encode_bit(s.has(f.name))

    for i in range(reserved):
      self.encode_bit(False)

    for f in type.fields:
      if f.type != "bit" and s != None and s.has(f.name):
        self.encode(f.type, s.get(f.name))

    self.flush()

  def decode_struct_body(self, type):
    reserved = 8*type.pack - len(type.fields)
    assert reserved >= 0

    s = qpid.Struct(type)

    for f in type.fields:
      if f.type == "bit":
        s.set(f.name, self.decode_bit())
      elif self.decode_bit():
        s.set(f.name, None)

    for i in range(reserved):
      if self.decode_bit():
        raise ValueError("expecting reserved flag")

    for f in type.fields:
      if f.type != "bit" and s.has(f.name):
        s.set(f.name, self.decode(f.type))

    self.clearbits()

    return s

  def encode_long_struct(self, s):
    enc = StringIO()
    codec = Codec(enc, self.spec)
    type = s.type
    codec.encode_short(type.type)
    codec.encode_struct_body(type, s)
    self.encode_longstr(enc.getvalue())

  def decode_long_struct(self):
    codec = Codec(StringIO(self.decode_longstr()), self.spec)
    type = self.spec.structs[codec.decode_short()]
    return codec.decode_struct_body(type)

  def decode_array(self):
    size = self.decode_long()
    code = self.decode_octet()
    count = self.decode_long()
    result = []
    for i in range(0, count):
      if self.types.has_key(code):
        value = self.decode(self.types[code])
      else:
        w = width(code)
        if fixed(code):
          value = self.read(w)
        else:
          value = self.read(self.dec_num(w))
      result.append(value)
    return result

def fixed(code):
  return (code >> 6) != 2

def width(code):
  # decimal
  if code >= 192:
    decsel = (code >> 4) & 3
    if decsel == 0:
      return 5
    elif decsel == 1:
      return 9
    elif decsel == 3:
      return 0
    else:
      raise ValueError(code)
  # variable width
  elif code < 192 and code >= 128:
    lenlen = (code >> 4) & 3
    if lenlen == 3: raise ValueError(code)
    return 2 ** lenlen
  # fixed width
  else:
    return (code >> 4) & 7
